### DISCUSSION TITLE
articles/230617-low-cost-k8s

### DISCUSSION THREAD

**Discussion by giscus** [zifeo/zifeo#1] [Comments]
# articles/230617-low-cost-k8s

How to deploy k8s on a $30.— monthly budget or less.

https://zifeo.com/articles/230617-low-cost-k8s

**Comment by bernardhalas** [comment]
Thanks for the nice write-up. Indeed, RKE2 is an option.

For those, who prefer Infrastructure-as-Code and GitOps compatible tools, I'd recommend Claudie though. Moreover, it's able to run cluster across on-prem and multiple cloud providers simultaneously and for long-term (in case you're running out of your on-prem capacity for any reason). www.github.com/berops/claudie

I'm one of the authors of Claudie

**Reply by zifeo** [reply]
@bernardhalas thanks for pointing out at other interesting options, do you provide any case study/stability benchmark (e.g. network-cut between providers and node pools)?

**Reply by bernardhalas** [reply]
Thank you for the question @zifeo. We've benchmarked these aspects so far.

1. The impact of traffic encryption on throughput, latency, and cpu overhead [here](https://www.berops.com/traffic-encryption-performance-in-kubernetes-clusters/)
2. The effect of latency (coming from geographical distribution for example) on control plane performance, more specifically on etcd [here](https://www.berops.com/evaluating-etcds-performance-in-multi-cloud/)

If you have a special interest in any other aspects, we're keen to know.

**Comment by mhazan01** [comment]
Hello,

i've deployed my own openstack 2023.1 on ubuntu 22 with kolla-ansible and ceph storage. i've been trying to deploy rke2 following your guide, however i get this error during the apply, i feel i'm missing a service on openstack or a config line in main but i'm not sure what, do you have any idea ? it success fully create one instance : 	single-server-agent-a-1 and it's volumes (so it can access cinder ?).

│ Error: error creating OpenStack object storage client: No suitable endpoint could be found in the service catalog.
│
│   with module.rke2.openstack_objectstorage_container_v1.etcd_snapshots[0],
│   on .terraform/modules/rke2/s3.tf line 1, in resource "openstack_objectstorage_container_v1" "etcd_snapshots":
│    1: resource "openstack_objectstorage_container_v1" "etcd_snapshots" {
│
╵
╷
│ Error: error creating OpenStack object storage client: No suitable endpoint could be found in the service catalog.
│
│   with module.rke2.openstack_objectstorage_container_v1.restic,
│   on .terraform/modules/rke2/s3.tf line 8, in resource "openstack_objectstorage_container_v1" "restic":
│    8: resource "openstack_objectstorage_container_v1" "restic" {
│
╵
╷
│ Error: error creating OpenStack object storage client: No suitable endpoint could be found in the service catalog.
│
│   with module.rke2.openstack_objectstorage_container_v1.velero,
│   on .terraform/modules/rke2/s3.tf line 13, in resource "openstack_objectstorage_container_v1" "velero":
│   13: resource "openstack_objectstorage_container_v1" "velero" {
│
Thanks for any idea :)

**Reply by zifeo** [reply]
@mhazan01 This is definitely related to OpenStack, which I sadly do not know very well. Maybe [check](https://github.com/hashicorp/terraform/issues/1374#issuecomment-89995418) the region or storage zones?

**Reply by mhazan01** [reply]
thanks, i found the issue which was indeed linked to openstack, the endpoint where configured with (projects_id) instead of (tenant_id), i guess i followed the wrong doc.

i also found another issue with ubuntu 22.04, when using ssh-keygen, it generate a rsa keypair, but ubuntu doesn't support that anymore for remote auth, terraform is stuck on waiting i'm guessing the Executing: ["/bin/sh" "-c" "      ssh-keygen -R hostname is broken ?

**Reply by zifeo** [reply]
You can disable this feature by setting `ff_write_kubeconfig = false`. The way the keygen is handled might be slightly different based on your local os and settings. You can then fetch manually your kubeconfig by SSHing into one of the server. Feel free to open an issue in the repo and/or submit a PR if you found a fix that works for you.

**Reply by mhazan01** [reply]
yeah now i'm facing another error with yq command, using --inplace instead of --in-place i think (testing pending)

**Reply by mhazan01** [reply]
i managed to finish deploying :)

need to generate and use stronger keypair and update main.tf (ie. "~/.ssh/id_ed25519.pub" )
the suggested boot_volume_size = 4 is not enough, just rke2 use at least 4.8gb.
i'm not seeing  rke2_volume being used ? no partition in fdisk -l on the node (but i might misunderstand it's uses)
i had a stupid dns issue (default is public dns) need to change dns_nameservers4 = [ "x.x.x.x","y.y.y.y"] in main.tf
need to change the "--inplace" for "--in-place -y" in .terraform/modules/rke2/keys.tf (might be due to debian yq version) and the sync fails but you get the data out in the local file.

thanks for the guide :)

**Reply by mhazan01** [reply]
and i'm back with another question :D now i'm trying to deploy rancher following the official doc. i don't understand how to deal with the ingress, if i understand right the loadbalancer we created on openstack replaces it ? but when i configure the port 443 in the loadbalancer and try to access the rancher ui, it doesn't seem to work. any suggestion ?

--edit : i found out that i have to setup another loadbalancer and deploy rancher with --set ingress.ingressClassName=nginx option set :)

**Reply by zifeo** [reply]
@mhazan01 you can also share/recycle the initial load balancer using this annotation:  `loadbalancer.openstack.org/load-balancer-id: "UUID of the load balancer"`
